### PR TITLE
WIP. Use venv for deps

### LINKED ISSUE #18: Doesn't work on distros with Python3 as default python version
Some distros (Archlinux for example) have Python 3 as default python version. So when install calls `pip` command, it will install dependencies in python 3 global environment. Also `/usr/bin/python` points to Python 3 binary.

Best fix would be to create virtualenv for srclib-python and install dependencies there. This would also benefit people with older python 2 as it would not litter global environment with extra packages and it would remove the need for `sudo`.

Example error:

```
Running: [/home/maiku/.srclib/sourcegraph.com/sourcegraph/srclib-python/.bin/srclib-python scan --repo github.com/MaikuMori/sourcegraph-atom --subdir .]
  --> with input {}
  File "/usr/bin/pydep-run.py", line 54
    print json.dumps(setup_infos)
             ^
SyntaxError: invalid syntax
EOF
```

### PR DESCRIPTION
**Pull request by MaikuMori**
This should fix #18.

`getVENVBinPath` is a bit misleading name as it will return `""` when ran inside docker container. I didn't want to complicate docker setup with virutalenv since it's pointless there.

### CODE REVIEW

**Comment by MaikuMori** [comment]
Currently it doesn't work fully when python project is using virtualenv (it doesn't scan the external dependencies and ignores them), but I'm quite sure it didn't work before as well.

**Comment by MaikuMori** [comment]
Now the tests are broken on the thing I actually got stuck on - getting refs to work.

File: python/util.go [hunk 1]
```diff
@@ -19,3 +22,14 @@ func runCmdStderr(cmd *exec.Cmd) error {
 	cmd.Stdout = os.Stderr
 	return cmd.Run()
 }
+
+func getVENVBinPath() (string, error) {
```
Comment (samertm) [inline python/util.go]: Can you add a comment to getVENVBinPath clarifying that it returns the empty string for Docker?

**Comment by MaikuMori** [comment]
Alright so part of the test fails because of this change https://github.com/sourcegraph/srclib/commit/98df7b1f9cac79d20b3562503df1d2eae864cfc5

I already fixed a related issue where `src` output didn't match what JavaScript toolchain expected.

Still working on tracking down the rest of the problems.

**Comment by MaikuMori** [comment]
This should fix toolchain installation on systems with Python 3 as default or when there are conflicting packages in systems environment (like Pip 6.0, sourcegraph/pydep#8).

But there is bigger problem with program mode. It assumes that package that is being analyzed is installed in the virtualenv (before system env). And obviously it's not installed in toolchains virtualenv, but it shouldn't be installed globally either (as required before).

So the way I see it there are couple ways to fix it:

 - Do the same thing as in docker mode, install package in toolchains virtualenv. Problem is that if user has multiple python projects it will turn into garbage bin with possible conflicts between packages, constant package reinstallation and other unforeseeable bugs. This is probably the simples solution.
 - Try to figure out which virtualenv is active from environment variables and use that (or systems if no virtualenv is active). I sorta like this solution except I wouldn't want to pollute it with grapher python package (to my understanding it has to be installed and run in same environment as package which is analyzed and it's deps). Another bonus here is that python project deps are quite often out of sync with requirements.txt while you're developing, this would pickup any new installed dependency instantly)
 - Make new virtual env in say `.srclib-cache` dir (or the new store @samertm is working on) install package/deps + grapher + pydep in that virtualenv and run from there. Maybe the best solution at cost of setup time (install packages, keep the in sync with changes in requirements.txt)

**Comment by samertm** [comment]
On Mon, Mar 16, 2015 at 3:03 PM, Miks Kalnins <notifications@github.com>
wrote:

>
>    - Do the same thing as in docker mode, install package in toolchains
>    virtualenv. Problem is that if user has multiple python projects it will
>    turn into garbage bin with possible conflicts between packages, constant
>    package reinstallation and other unforeseeable bugs. This is probably the
>    simples solution.
>
> I don't think we should do this.

>
>    - Try to figure out which virtualenv is active from environment
>    variables and use that (or systems if no virtualenv is active). I sorta
>    like this solution except I wouldn't want to pollute it with grapher python
>    package (to my understanding it has to be installed and run in same
>    environment as package which is analyzed and it's deps). Another bonus here
>    is that python project deps are quite often out of sync with
>    requirements.txt while you're developing, this would pickup any new
>    installed dependency instantly)
>
> I think the best case scenario would be if we could get this to work, but
I don't know much about virtualenv. Is the issue that srclib-python and the
app it's analyzing both need their *own* virtualenvs? We don't want to
pollute the user's requirements.txt or their other virtualenv configs.

>
>    - Make new virtual env in say .srclib-cache dir (or the new store
>    @samertm <https://github.com/samertm> is working on) install
>    package/deps + grapher + pydep in that virtualenv and run from there. Maybe
>    the best solution at cost of setup time (install packages, keep the in sync
>    with changes in requirements.txt)
>
> If we can't do the second point, I think this is the second best. For this
to work, the should be a command like "src toolchain temp-dir
<toolchain-path>" that gives you back a full path to a temp directory that
you could write to, like "/home/<user>/.srclib/tmp/<toolchain-path>/", and
then srclib-python would create the virtualenv it needs to analyze the app.

What do you think?

**Comment by samertm** [comment]
Actually, I just thought of something. If we only require that a python program has a requirements.txt file, then it makes more sense to build the virtual env ourselves rather than look for the virtual env the user created or require that srclib-python be run in their virtualenv. Does that make sense?

**Comment by MaikuMori** [comment]
> Does that make sense?

That's the option number 3. It works.

> Is the issue that srclib-python and the app it's analyzing both need their *own* virtualenvs?

Neither should be putting stuff in systems one. If we build venv for app using thier requirements.txt, we can also add our own stuff there. I suppose we would still need 1 virtual env for toolchain because in order to find all deps you need to run pydep which needs to be installed somewhere. Also venvs allow symlinking which we can take advantage of.

Abstract representation:

```
~/.srclib/sourcegraph.com/sourcegraph/srclib-python/.env
  - pydep
  - grapher

~/.srclib/tmp/sourcegraph.com/sourcegraph/srclib-python/<project_path>/env
 - deps from requirements.txt/setup.py
 - symlink to pydep (maybe not needed)
 - symlink to grapher
```

**Comment by samertm** [comment]
Yeah, the 3rd option makes the most sense. If you want to start working on it, you can just assume that "~/.srclib/tmp/sourcegraph.com/sourcegraph/srclib-python" exists, and I can ping you when I've added that to srclib. (If you want to add the "toolchain tmp" command, I can show you where to start.)

**Pull request merged**